website/src/documentation/programming-guide.md

```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
+When a `DoFn` contains multiple state specifications, reading each one in order can be slow. Calling the read() function
+on a state can cause the runner to perform a blocking read. Performing multiple blocking reads in sequence adds latency
+to element processing. The readLater method allows the runner to know that the state will be read in the future, 
+allowing multiple state reads to be batched together.
+```java
+PCollection<KV<String, ValueT>> perUser = readPerUser();
+perUser.apply(ParDo.of(new DoFn<KV<String, ValueT>, OutputT>() {
+  @StateId("state1") private final StateSpec<ValueState<Integer>> state1 = StateSpecs.value();
+  @StateId("state2") private final StateSpec<ValueState<String>> state2 = StateSpecs.value();
+  @StateId("state3") private final StateSpec<BagState<ValueT>> state3 = StateSpecs.bag();
+
+  @ProcessElement public void process(
+    @StateId("state1") ValueState<Integer> state1,
+    @StateId("state2") ValueState<String> state2,
+    @StateId("state3") BagState<ValueT> state3) {
+    state1.readLater();
+    state2.readLater();
+    state3.readLater();
+   
+    // The runner can now batch all three states into a single read, reducing latency.
+    processState1(state1.read());
+    processState2(state2.read());
+    processState3(state3.read());
+  }
+}));
+```
+
+### 10.3 Timers {#timers}
+Beam provides a per-key timer callback API. This allows for delayed processing of data aggregated using the state API.
+Timers can be set to callback at either an event-time or a processing-time timestamp. Every timer is identified with a
+TimerId. A given timer for a keycan only be set for a single timestamp. Calling set on a timer overwrites the previous
+firing time for that key's timer.
+
+#### 10.3.1 Event-time timers {#event-time-timers}
+Event-time timers fire when the input watermark for the DoFn passes the time at which the timer is set, meaning that 
+the runner believes that there are no more elements to be processed with timestamps before the timer timestamp. This
+allows for event-time aggregations. 
+
+```java
+PCollection<KV<String, ValueT>> perUser = readPerUser();
+perUser.apply(ParDo.of(new DoFn<KV<String, ValueT>, OutputT>() {
+  @StateId("state") private final StateSpec<ValueState<Integer>> state = StateSpecs.value();
+  @TimerId("timer") private final TimerSpec timer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+  @ProcessElement public void process(
+      @Element KV<String, ValueT> element,
+      @Timestamp Instant elementTs,
+      @StateId("state") ValueState<Integer> state, 
+      @TimerId("timer") Timer timer) {
+     ...
+     // Set an event-time timer to the element timestamp.
+     timer.set(elementTs);
+  }
+  
+   @OnTimer("timer") public void onTimer() {
+      //Process timer.
+   }
+}));
+
+```
+#### 10.3.2 Processing-time timers {#processing-time-timers}
+Processing-time timers fire when the real wall-clock time passes. This is often used to create larger batches of data
```

Comment:
Worth mentioning expected behaviour if used in batch pipeline. If I recall direct runner doesnt fire process time timers?